File: clients/python/model/level.py
from .tile import Tile
class Level:
    def __init__(self, tiles):
        self.tiles = tiles
    @staticmethod
    def read_from(stream):
        tiles = []
        for _ in range(stream.read_int()):
            tiles_element = []
            for _ in range(stream.read_int()):
                tiles_element_element = Tile(stream.read_int())
                tiles_element.append(tiles_element_element)
            tiles.append(tiles_element)
        return Level(tiles)
    def write_to(self, stream):
        stream.write_int(len(self.tiles))
        for element in self.tiles:
            stream.write_int(len(element))
            for element in element:
                stream.write_int(element)
    def __repr__(self):
        return "Level(" + \
            repr(self.tiles) + \
            ")"

File: clients/cpp/model/Level.cpp
#include "Level.hpp"

Level::Level() { }
Level::Level(std::vector<std::vector<Tile>> tiles) : tiles(tiles) { }
Level Level::readFrom(InputStream& stream) {
    Level result;
    result.tiles = std::vector<std::vector<Tile>>(stream.readInt());
    for (size_t i = 0; i < result.tiles.size(); i++) {
        result.tiles[i] = std::vector<Tile>(stream.readInt());
        for (size_t j = 0; j < result.tiles[i].size(); j++) {
            switch (stream.readInt()) {
            case 0:
                result.tiles[i][j] = Tile::EMPTY;
                break;
            case 1:
                result.tiles[i][j] = Tile::WALL;
                break;
            case 2:
                result.tiles[i][j] = Tile::PLATFORM;
                break;
            case 3:
                result.tiles[i][j] = Tile::LADDER;
                break;
            case 4:
                result.tiles[i][j] = Tile::JUMP_PAD;
                break;
            default:
                throw std::runtime_error("Unexpected discriminant value");
            }
        }
    }
    return result;
}
void Level::writeTo(OutputStream& stream) const {
    stream.write((int)(tiles.size()));
    for (const std::vector<Tile>& tilesElement : tiles) {
        stream.write((int)(tilesElement.size()));
        for (const Tile& tilesElementElement : tilesElement) {
            stream.write((int)(tilesElementElement));
        }
    }
}
std::string Level::toString() const {
    return std::string("Level") + "(" +
        "TODO" + 
        ")";
}
